### PR TITLE
CreatePromptlabRun API

### PR DESCRIPTION
**Pull request by prithvikannan**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Add a public undocumented `runs/create-promptlab-run` API to power prompt engineering UI. 

In this PR, we do the following
* Create `promptlab_utils.py` to handle logic for generating prompt engineering files
* Create `_promptlab.py` with a private model flavor and loader module
* Extend filestore to support `create_promptlab_run` 
* Extend sqlalchemy store to support `create_promptlab_run` 

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

This PR is supported by the following tests
* Added tests in `test_rest_tracking` for hitting the `runs/create-promptlab-run` API
* Added tests in `test_file_store` for `_create_promptlab_run` impl
* Added tests in `test_sqlalchemy_store` for `_create_promptlab_run` impl
* Added tests in `test_promptlab_utils` for promptlab file tests

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/promptlab_utils.py
Comment (harupy) [inline mlflow/utils/promptlab_utils.py:17]:
```suggestion
    conda_yaml = f"""
name: mlflow-env
channels:
  - defaults
dependencies:
  - pip
  - pip:
    - mlflow[gateway]=={mlflow_version}
"""
```

Comment (harupy) [reply mlflow/utils/promptlab_utils.py:17]: Can we fix the channel to `conda-forge`?

File: mlflow/utils/promptlab_utils.py
Comment (harupy) [inline mlflow/utils/promptlab_utils.py:97]:
```suggestion
    loader_module_text = f'''
```

to use `"""` in the string.

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (harupy) [inline mlflow/store/tracking/sqlalchemy_store.py:1064]: /ai join this using `*` like `[*a, b, c]`.

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (harupy) [inline mlflow/store/tracking/sqlalchemy_store.py:1090]: Can we remove this directory once logging is complete?

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (harupy) [inline mlflow/store/tracking/sqlalchemy_store.py:1081]: /ai Should we use `utcnow` here if we want UTC time?

File: mlflow/utils/promptlab_utils.py
Comment (harupy) [inline mlflow/utils/promptlab_utils.py:17]: /ai In this python code, I think we can use `textwrap.dedent` to remove trailing space on each line. Is it slow?

File: mlflow/utils/promptlab_utils.py
Comment (harupy) [inline mlflow/utils/promptlab_utils.py:113]: Can we create a template as a python file, read it, and replace `__xx___`?

```python
# loader/gateway_loader_module.py
from typing import List, Dict
from string import Template
import pandas as pd
import mlflow.gateway

mlflow.gateway.set_gateway_uri("databricks")


class GatewayModel:
    def __init__(self, model_path):
        self.prompt_template = Template(
            """
__sanitized_python_template__
"""
        )

    def predict(self, inputs: pd.DataFrame) -> List[str]:
        results = []
        for idx in inputs.index:
            prompt = self.prompt_template.substitute(__python_inputs__)
            result = mlflow.gateway.query(
                route="{model_route}",
                data={
                    {
                        "prompt": prompt,
                        **__python_parameters__,
                    }
                },
            )
            results.append(result["candidates"][0]["text"])
        return results


def _load_pyfunc(model_path):
    return GatewayModel(model_path)

```

File: mlflow/utils/promptlab_utils.py
Comment (prithvikannan) [reply mlflow/utils/promptlab_utils.py:97]: If we make this change as you propose, won't we have the same problem if we try to use `'''` in the contents?

We have logic to escape the `"""` sequence so it doesn't show up in the contents. Do you think that is enough?

Comment (harupy) [reply mlflow/utils/promptlab_utils.py:97]: Yes, the same problem will occur for `'''`. I just wanted to change:


```python
"""
    def __init__(self, model_path):
        self.prompt_template = Template(\"\""
{sanitized_python_template}\n\""")
"""
```

to:

```python
'''
    def __init__(self, model_path):
        self.prompt_template = Template("""
{sanitized_python_template}\n""")
'''
```

File: mlflow/utils/promptlab_utils.py
Comment (prithvikannan) [reply mlflow/utils/promptlab_utils.py]: Yea I think this is a great idea. I push commits that use a template txt file. I am not sure how to make a python file with these template variables.

File: mlflow/utils/promptlab_utils.py
Comment (prithvikannan) [reply mlflow/utils/promptlab_utils.py]: Gotcha. The templated file solves this problem anyway.

File: mlflow/store/tracking/file_store.py
Comment (harupy) [inline mlflow/store/tracking/file_store.py:1391]: /ai Explain why we should use `with tempfile.TemporaryDirectory():` instead of `tempfile.mkdtemp()` and `shutil.rmtree(local_dir)`.

File: mlflow/utils/promptlab_utils.py
Comment (harupy) [inline mlflow/utils/promptlab_utils.py:87]: curious why we need to replace `{{}}`.

File: mlflow/utils/promptlab_utils.py
Comment (harupy) [inline mlflow/utils/promptlab_utils.py:90]: Is this line needed?

File: mlflow/utils/promptlab_utils.py
Comment (harupy) [reply mlflow/utils/promptlab_utils.py:87]: Any chance that `prompt_template` contains `{{` or `}}` that shouldn't be replaced?

File: mlflow/utils/promptlab_utils.py
Comment (harupy) [inline mlflow/utils/promptlab_utils.py:96]: We can use `__file__` here. 
```suggestion
    template_path = Path(__file__).parent.joinpath("promptlab_loader_module_template.txt")
    with template_path.open() as loader_module_template:
```

File: mlflow/utils/promptlab_utils.py
Comment (harupy) [inline mlflow/utils/promptlab_utils.py:103]: This approach seems risky. We can easily inject malicious code.

Comment (harupy) [reply mlflow/utils/promptlab_utils.py:103]: I see the template is wrapped by `"""`.

File: mlflow/utils/promptlab_utils.py
Comment (harupy) [inline mlflow/utils/promptlab_utils.py:46]: Can we create this dict using `mlflow.models.model.Model`?

File: mlflow/utils/promptlab_loader_module_template.txt
Comment (harupy) [inline mlflow/utils/promptlab_loader_module_template.txt:34]: Can we add a `prompt_lab` flavor (private module)?

File: mlflow/utils/promptlab_utils.py
Comment (prithvikannan) [reply mlflow/utils/promptlab_utils.py:87]: Python Template string https://www.geeksforgeeks.org/template-class-in-python use `$` before template vars.

```
# A Simple Python template example
from string import Template
  
# Create a template that has placeholder for value of x
t = Template('x is $x')
  
# Substitute value of x in above template
print (t.substitute({'x' : 1}))
```

Comment (prithvikannan) [reply mlflow/utils/promptlab_utils.py:87]: > Any chance that prompt_template contains {{ or }} that shouldn't be replaced?

double braces in `prompt_template` are treated as input variables in the promptlab UI, so i dont think this is possible. the frontend create the realized prompt by filling in the var content into the vars denoted by double braces and make the call to gateway.

File: mlflow/utils/promptlab_utils.py
Comment (prithvikannan) [reply mlflow/utils/promptlab_utils.py:90]: Hmm, this is supposed to say `sanitized_python_template = python_template.replace('"""', '\"\"\"')` but black formatter removes the escaped quote. Added a `# fmt: off`/`# fmt: on` block to get around this.

File: mlflow/utils/promptlab_loader_module_template.txt
Comment (prithvikannan) [reply mlflow/utils/promptlab_loader_module_template.txt:34]: will defer to @dbczumar

File: mlflow/utils/promptlab_utils.py
Comment (prithvikannan) [reply mlflow/utils/promptlab_utils.py:103]: I agree that this code is vulnerable to this sort of prompt injection. However, regardless of promptlab, a user is free to construct a model such that `predict` "steals secrets".

I am not sure if we need to change our security posture for these models that are generated by the promptlab.

Comment (dbczumar) [reply mlflow/utils/promptlab_utils.py:103]: Agreed!

File: mlflow/utils/promptlab_loader_module_template.txt
Comment (dbczumar) [reply mlflow/utils/promptlab_loader_module_template.txt:34]: This is a good idea. This way, we can remove the loader module code from the scala backend entirely. The scala backend can reference "mlflow._promptlab"

File: mlflow/utils/promptlab_utils.py
Comment (harupy) [reply mlflow/utils/promptlab_utils.py:92]: /ai Are there any other options?

Comment (harupy) [reply mlflow/utils/promptlab_utils.py:92]: Can we use Jinja?

https://realpython.com/primer-on-jinja-templating/

File: mlflow/utils/promptlab_utils.py
Comment (harupy) [reply mlflow/utils/promptlab_utils.py:111]: Got it, I agree. Can we use files instead of constructing code?

File: mlflow/utils/promptlab_utils.py
Comment (harupy) [reply mlflow/utils/promptlab_utils.py:92]: /ai I have a string that looks like a Jinja's template as shown below:

```
Hello, {{ name }}.
```

Give me code that that converts this to this using regular expression:

```
Hello, $name.
```

File: mlflow/_promptlab.py
Comment (dbczumar) [inline mlflow/_promptlab.py:93]:
```suggestion
        loader_module="mlflow._promptlab",
```

File: mlflow/_promptlab.py
Comment (dbczumar) [inline mlflow/_promptlab.py:132]: I don't think we should be pickling an instance of `_PromptlabModel`. Instead, `save_model()` should take in   arguments like `prompt`, `temperature`, `max_tokens`, etc., and we should construct an instance of a pyfunc-compatible model wrapper (the same way your loader module code from Scala does) at load time

File: mlflow/_promptlab.py
Comment (dbczumar) [inline mlflow/_promptlab.py:138]: Neither of these deps are necessary. Are any dependencies necessary beyond just `mlflow`? If not, we can remove this method

File: mlflow/_promptlab.py
Comment (harupy) [inline mlflow/_promptlab.py:1]: Is this line necessary?

File: mlflow/_promptlab.py
Comment (harupy) [inline mlflow/_promptlab.py:55]: /ai what value is passed to the `data` argument in this function call?

File: mlflow/_promptlab.py
Comment (harupy) [inline mlflow/_promptlab.py:44]:
```suggestion
                param.key: inputs[f"{param.key}"][idx] for param in self.prompt_parameters
```
Is `f` missing?

File: mlflow/_promptlab.py
Comment (harupy) [reply mlflow/_promptlab.py:55]: `dict.update` returns `None`. `{...}` is a set literal. `{None,}` is passed to `data`:

```python
prompt = 1
model_parameters_as_dict = {}
data = {
    {
        "prompt": prompt,
    }.update(model_parameters_as_dict),
}
print(data)
# {None}
```

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (harupy) [inline mlflow/store/tracking/sqlalchemy_store.py:1120]: Is this line correct? `int(datetime.now().strftime("%Y%m%d"))` returns an integer like `20230831`.

File: mlflow/store/tracking/file_store.py
Comment (harupy) [inline mlflow/store/tracking/file_store.py:1338]: When do we fail to import these?

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (harupy) [inline mlflow/store/tracking/sqlalchemy_store.py:1066]: Can we log tags with `self.log_batch`?

File: mlflow/utils/promptlab_utils.py
Comment (harupy) [inline mlflow/utils/promptlab_utils.py:4]:
```suggestion
def create_eval_results_json(prompt_parameters, model_input, model_output_parameters, model_output):
```

`create_eval_results_json` seems more accurate.

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (prithvikannan) [reply mlflow/store/tracking/sqlalchemy_store.py:1120]: good catch. this should be `int(time.time() * 1000)`.

on this the RunInfo docs https://mlflow.org/docs/latest/python_api/mlflow.entities.html?highlight=end_time#mlflow.entities.RunInfo.end_time
> End time of the run, in number of milliseconds since the UNIX epoch.

File: mlflow/_promptlab.py
Comment (prithvikannan) [inline mlflow/_promptlab.py:133]: i think this should be `2.7.0` but how can we handle this before 2.7 is released @dbczumar @harupy

Comment (harupy) [reply mlflow/_promptlab.py:133]: We're in MLflow. Can we use `mlflow.__version__`?

File: mlflow/_promptlab.py
Comment (harupy) [inline mlflow/_promptlab.py:23]:
```suggestion
                param.key: inputs[param.key][idx] for param in self.prompt_parameters
```

**Review by harupy — APPROVED** [review]
left a few more comments, otherwise LGTM!

File: mlflow/store/tracking/file_store.py
Comment (harupy) [inline mlflow/store/tracking/file_store.py:1312]: /ai join this using `*`

File: mlflow/store/tracking/file_store.py
Comment (harupy) [reply mlflow/store/tracking/file_store.py:1339]: Got it!

File: mlflow/store/tracking/file_store.py
Comment (harupy) [inline mlflow/store/tracking/file_store.py:1348]:
```suggestion

        except ImportError:
            signature = None
        else:
            inputs_colspecs = [ColSpec(DataType.string, param.key) for param in prompt_parameters]
            outputs_colspecs = [ColSpec(DataType.string, "output")]
            signature = ModelSignature(
                inputs=Schema(inputs_colspecs),
                outputs=Schema(outputs_colspecs),
            )
```

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (harupy) [inline mlflow/store/tracking/sqlalchemy_store.py:1120]: should we mark the run as `FAILED` if we eucounter an unexpected failure?

File: mlflow/utils/promptlab_loader_module_template.txt
Comment (harupy) [inline mlflow/utils/promptlab_loader_module_template.txt:1]: Is this file still used?

File: mlflow/store/tracking/file_store.py
Comment (harupy) [reply mlflow/store/tracking/file_store.py:1312]: /ai I meant something like this `[*a, b, c]`.

Comment (harupy) [reply mlflow/store/tracking/file_store.py:1312]:
```suggestion
        parameters_to_log = [*model_parameters, Param("model_route", model_route), Param("prompt_template", prompt_template)]
```

for consistency

File: mlflow/_promptlab.py
Comment (prithvikannan) [reply mlflow/_promptlab.py:133]: we could probably, but i am not sure if that is desired behavior. for example, once we release 2.7.1, we do not enforce that the promptlab model has 2.7.0, just that it has 2.7.0.

File: mlflow/_promptlab.py
Comment (dbczumar) [inline mlflow/_promptlab.py:136]: @prithvikannan I think the MLflow version gets added as a requirement automatically in `_process_pip_requirements`, so we can remove this and just pass an empty list.

Ref: https://github.com/mlflow/mlflow/blob/37298ffafcd34002352d01d579d4524790544267/mlflow/utils/environment.py#L494-L495

File: mlflow/store/tracking/file_store.py
Comment (dbczumar) [inline mlflow/store/tracking/file_store.py:1302]: Nit: Can we update the docstring here to explain what promptlab run creation is?

File: mlflow/server/handlers.py
Comment (dbczumar) [inline mlflow/server/handlers.py:1224]: @prithvikannan Do we strictly need separate `_create_promptlab_run` implementations in file_store.py and sqlalchemy_store.py? The only difference between the two impls seems to be how we fetch the artifact location, but we can get this info from `get_run()`. Can we define a single method in `promptlab_utils.py` that takes a store as a param and does all the logging?

**Review by dbczumar — APPROVED** [review]
@prithvikannan LGTM once comments are addressed. Thanks so much!

**Pull request merged**
